### PR TITLE
Done with squashing

### PR DESCRIPTION
**Pull request by janehyt**
1. add some methods which related to authorize_security_group and revoke_security_group

2. complete security_group_rule model

Something fogot to push last time..related ti security group egress rule.

### CODE REVIEW

**Comment by janehyt** [comment]
@plribeiro3000 please check with this.

**Comment by plribeiro3000** [comment]
Sure, i will. =)

File: lib/fog/aliyun/models/compute/security_group.rb [hunk 1]
```diff
@@ -4,82 +4,83 @@ module Fog
...
+        attribute :name,         :aliases => "SecurityGroupName"
+        attribute :description,  :aliases => "Description"
+        attribute :created_at,   :aliases => "CreationTime"
+        attribute :vpc_id,       :aliases => "VpcId"
 
         def vpc
-            requires :vpc_id
-            $vpc=Fog::Compute::Aliyun::Vpcs.new(:service=>service).all('vpcId'=>vpc_id)[0]
+          requires :vpc_id
+          $vpc=Fog::Compute::Aliyun::Vpcs.new(:service=>service).all('vpcId'=>vpc_id)[0]
```
Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group.rb]: Why are you defining a global variable here?
Do you want to memoize the response? If so use `@vpc ||=` instead of `$vpc`.

File: lib/fog/aliyun/models/compute/security_group.rb [hunk 2]
```diff
@@ -4,82 +4,83 @@ module Fog
...
-            options[:vpcId] = vpc_id if vpc_id
-            options[:name] = name if name
-            options[:description] = description if description
-            data = Fog::JSON.decode(service.create_security_group(options).body)
-            true
+          options[:vpcId] = vpc_id if vpc_id
+          options[:name] = name if name
+          options[:description] = description if description
+          data = Fog::JSON.decode(service.create_security_group(options).body)
+          true
```
Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group.rb]: This method will always return true. What about a failure to save the object?

File: lib/fog/aliyun/models/compute/security_group.rb [hunk 3]
```diff
@@ -4,82 +4,83 @@ module Fog
...
-        #   attributes[:security_group_rules] = new_rules
-        # end
-
-        # def save
-        #   requires :name, :description
-        #   data = service.create_security_group(name, description)
-        #   merge_attributes(data.body['security_group'])
-        #   true
-        # end
+        def authorize_seurity_group_sg_rule(group_id,direction="ingress",options={})
```
Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group.rb]: misspelling. =)
`authorize_seurity_group_sg_rule` should be `authorize_security_group_sg_rule`, no?

**Comment by plribeiro3000** [comment]
@janehyt It seems the code changed a little bit otherwise i might have missed some details.
All methods are always returning true. Do you think it would be harder to check the result of the request?

File: lib/fog/aliyun/models/compute/security_group.rb (hunk 2 above)
Comment (janehyt) [inline lib/fog/aliyun/models/compute/security_group.rb]: Actually, if a failure occurs, the methods will throw an exception. I return true just so that my pg could be consistent with other models.

File: lib/fog/aliyun/models/compute/security_group.rb (hunk 3 above)
Comment (janehyt) [inline lib/fog/aliyun/models/compute/security_group.rb]: yes,I've fix it.

File: lib/fog/aliyun/models/compute/security_group.rb (hunk 2 above)
Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group.rb]: so the behaviour is:

* true if worked
* exception if failed

I'm not sure about this but lets move on with this.

File: lib/fog/aliyun/models/compute/security_group.rb [hunk 4]
```diff
@@ -4,82 +4,86 @@ module Fog
...
+        attribute :vpc_id,       :aliases => "VpcId"
 
         def vpc
-            requires :vpc_id
-            $vpc=Fog::Compute::Aliyun::Vpcs.new(:service=>service).all('vpcId'=>vpc_id)[0]
+          requires :vpc_id
+          @vpc||= begin
+            Fog::Compute::Aliyun::Vpcs.new(
+              :service=>service).all('vpcId'=>vpc_id)[0]
+          end
```
Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group.rb]: You can call the method `vcps` on the `service` method and achieve the same result. And instead of using the collection method you could use the getter one.

```ruby
service.vpcs. get(vpc_id)
```

**Comment by plribeiro3000** [comment]
@janehyt BTW, you don't need to open a new PR after squashing. You can do a `git push origin --force` and it will rewrite the remote branch and github with automagically display the changes including the removal of commits. =)

File: lib/fog/aliyun/models/compute/security_group.rb (hunk 1 above)
Comment (janehyt) [inline lib/fog/aliyun/models/compute/security_group.rb]: I find that I forgot to click the commet yesterday...and should I change vpc method in other models? Shoud I use @vpc ||= instead of $vpc in the get method of vpcs?

Comment (janehyt) [inline lib/fog/aliyun/models/compute/security_group.rb]: It seems like I forgot to clickc comment button yesterday. Define a global variable just because when save vswitch, it need the vpc. And should I change  vpc methods in other models? Should I fix the $vpc in vpcs collection?

Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group.rb]: Sure @janehyt. :+1:

**Comment by janehyt** [comment]
@plribeiro3000  thanks. Related to the global variable vpc, it seems like I forgot to clickc comment button yesterday. Define a global variable just because when save vswitch, it need the vpc. And should I change vpc methods in other models? Should I fix the $vpc in vpcs collection?

File: lib/fog/aliyun/models/compute/security_group.rb (hunk 1 above)
Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group.rb]: That global variable will bit your ass if you have 2 of the same object so we need to change every place is using it. We can do it in another PR if it is code you did not change.

**Comment by plribeiro3000** [comment]
@janehyt Yes. That is going to bite some asses if the use has 2 instances of this object.
We can take care of it in another PR if you think its better.

**Comment by janehyt** [comment]
@plribeiro3000 yes, you're right. So there's nothing to change now?

**Comment by plribeiro3000** [comment]
Let me see.

File: lib/fog/aliyun/models/compute/security_group.rb [hunk 5]
```diff
@@ -4,82 +4,84 @@ module Fog
...
+          requires :vpc_id
+          service.vpcs.get(vpc_id)
+          end
         end
 
         def security_group_rules
-            requires :id
-            Fog::Compute::Aliyun::SecurityGroupRules.new(:service=>service).get(id)
+          requires :id
+          Fog::Compute::Aliyun::SecurityGroupRules.new(:service=>service).get(id)
```
Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group.rb]: i think you could use `service.security_group_roles.get(id)` here.

File: lib/fog/aliyun/models/compute/security_group.rb [hunk 6]
```diff
@@ -4,82 +4,84 @@ module Fog
...
         def save(options={})
-            options[:vpcId] = vpc_id if vpc_id
-            options[:name] = name if name
-            options[:description] = description if description
-            data = Fog::JSON.decode(service.create_security_group(options).body)
-            true
+          options[:vpcId] = vpc_id if vpc_id
+          options[:name] = name if name
+          options[:description] = description if description
+          data = Fog::JSON.decode(service.create_security_group(options).body)
```
Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group.rb]: This encoding and decoding should be up in the request itself. The model should work on top of a ruby hash and deal with business logic only.

Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group.rb]: And you don't need to save the return of it in an attribute `data` because your return will always be `true`.

File: lib/fog/aliyun/models/compute/security_group.rb [hunk 7]
```diff
@@ -4,82 +4,84 @@ module Fog
...
-        #   Fog::Logger.deprecation('#create_security_group_rule is deprecated. Use the Fog::Compute::Openstack::SecurityGroupRule objects to destroy rules.')
-        #   service.delete_security_group_rule(rule_id)
-        #   true
-        # end
+        def revoke_security_group_ip_rule(cidr_ip,direction="ingress",options={})
+          requires :id
+          nic_type=options[:nic_type]
+          unless nic_type
+            nic_type="internet"
+          end
```
Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group.rb]: here you could use `nic_type = options.fetch(:nic_type, "internet")`. If it can't find the key it will use the second parameter as the default value.

File: lib/fog/aliyun/models/compute/security_group_rule.rb [hunk 8]
```diff
@@ -19,22 +19,28 @@ class SecurityGroupRule < Fog::Model
...
+              service.create_security_group_sg_rule(security_group_id,source_group_id,options)
+            end
           end
         end
-
-        def destroy
-          requires :id
-          service.delete_security_group_rule(id)
-          true
-        end
```
Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group_rule.rb]: Why did you remove this method?

**Comment by plribeiro3000** [comment]
@janehyt Did some other small comments.
Regarding the encoding and decoding i just realized you are just following the current conventions so not needed to fix it in this PR.

File: lib/fog/aliyun/models/compute/security_group_rule.rb (hunk 8 above)
Comment (janehyt) [inline lib/fog/aliyun/models/compute/security_group_rule.rb]: Because delete security_group_rule is complecated, and you can refer to revoke_seurity_group_sg_rule,revoke_seurity_group_ig_rule

**Comment by janehyt** [comment]
@plribeiro3000 I've done with your comments.thanks

File: lib/fog/aliyun/models/compute/security_group_rule.rb (hunk 8 above)
Comment (plribeiro3000) [inline lib/fog/aliyun/models/compute/security_group_rule.rb]: Maybe you could leave the `destroy` method to maintain the same interface as the majority of `fog` but calling the service `revoke_seurity_group_ig_rule` instead. What do you think?

**Comment by plribeiro3000** [comment]
Sounds ok to me now. =)

:+1:

**Comment by janehyt** [comment]
@plribeiro3000  add the destroy now

**Comment by plribeiro3000** [comment]
Already merged. I guess you need to open another PR. =)

**Comment by janehyt** [comment]
ok, I see. =)
